### PR TITLE
Adds support for AWQ

### PR DESCRIPTION
**Pull request by JyotinderSingh**
This PR introduces support for AWQ (Activation-aware Weight Quantization) to Keras. AWQ is a post-training quantization technique specifically designed for Large Language Models (LLMs) that achieves significant model compression (4-bit integer weights) while maintaining high accuracy.

Unlike standard weight quantization which treats all weights equally, AWQ operates on the observation that protecting "salient" weights (those corresponding to larger activation magnitudes) is critical for performance. It performs a lightweight calibration to find optimal per-channel scaling factors that minimize quantization error.

Reference paper: [arXiv:2306.00978](https://arxiv.org/abs/2306.00978)

## Features
1. AWQConfig: A new configuration class `keras.quantizers.AWQConfig` to handle calibration datasets, tokenizer, group sizes, and grid search parameters.
2. Updates `Dense` and `EinsumDense` layers to support AWQ variable layouts (`quantized_kernel`, `awq_scales`, `kernel_scale`, `kernel_zero`, `g_idx`).
3. Algorithm: 
    - Collects activation statistics from a calibration dataset.
    - Performs grid search to find optimal scaling factors.
    - Applies scales to weights before quantization to preserve salient features.
    - Implements packed 4-bit storage.
4. Adds `AWQDTypePolicy` to handle policies like `"awq/4/128"`.

## API

To use AWQ, the user must provide a representative calibration dataset (e.g., a list of strings) and a tokenizer.

```py
import keras
from keras.quantizers import AWQConfig

# 1. Define configuration
awq_config = AWQConfig(
    dataset=["The quick brown fox...", "Another sentence..."], # Calibration data
    tokenizer=tokenizer,           # Tokenizer callable
    weight_bits=4,                 # Currently supports 4-bit
    group_size=128,                # Block size for quantization
    num_grid_points=20,            # Granularity of scale search
    
    # Optional: Define structure if model doesn't support `get_quantization_layer_structure` hook
    # quantization_layer_structure={
    #    "pre_block_layers": [model.layers[1]], 
    #    "sequential_blocks": [model.layers[2]]
    # }
)

# 2. Apply quantization
# This triggers the calibration loop (forward passes) and weight packing
model.quantize(config=awq_config)

# 3. Save the quantized model
model.save("awq_model.keras")
```

## Verification
1. Added unit tests covering the grid search algorithm, matrix quantization, and layer integration.
2. Integration Tests: End-to-end tests ensuring model.save/load works for AWQ models.
3. The test suite includes a Transformer accuracy test (`AWQAccuracyTest`) checking KL divergence and top-1 match rates on a small sequence classifier to ensure quantization preserves model behavior.

Accuracy Check: The test suite includes a Transformer accuracy test (AWQAccuracyTest) checking KL divergence and top-1 match rates on a small sequence classifier to ensure quantization preserves model behavior.

### Configuration

| Parameter | Value |
|-----------|-------|
| GPU | NVIDIA RTX 4070 Ti Super |
| Dataset | WikiText-2 |
| Sequence Length | 128 |
| Evaluation Batches | 50 |
| Calibration Samples | 128 |
| Grid Points | 20 |
| Runs per Model | 3 |

### Results

Source: [benchmarks.csv](https://github.com/user-attachments/files/24522118/awq_benchmarks.csv)

| Model | Pre PPL | Post PPL | PPL Δ | Disk Size (Pre→Post) | Disk Δ | GPU Mem (Pre→Post) | Mem Δ | Throughput (Pre→Post) | Throughput Δ | Quant Time |
|-------|---------|----------|-------|----------------------|--------|-------------------|-------|----------------------|--------------|------------|
| Qwen3 1.7B | 37.65 | 45.79 | +21.64% | 6.42→1.88 GB | -70.7% | 6.95→2.09 GB | -69.9% | 179.6→161.0 tok/s | -10.4% | 428s |
| Gemma3 1B | 172.45 | 178.03 | +3.23% | 3.73→1.48 GB | -60.2% | 3.88→1.62 GB | -58.3% | 399.5→337.7 tok/s | -15.5% | 225s |
| OPT 125M | 77.06 | 84.75 | +9.97% | 0.47→0.20 GB | -58.3% | 0.63→0.37 GB | -40.9% | 691.4→668.8 tok/s | -3.3% | 101s |

### Observations

- **Disk size reduction**: 58–71% across models (INT4 quantization from FP32/BF16 weights)
- **GPU memory reduction**: 41–70% during inference
- **Perplexity degradation**: +3.2% (Gemma3 1B) to +21.6% (Qwen3 1.7B)
- **Throughput**: -10% to -15% for larger models; OPT 125M shows -3% average with high variance
- **Quantization time**: Scales linearly with model size (~101s for 125M, ~428s for 1.7B)

### Reproduction
```bash
docker run --gpus all \
    -v $(pwd)/outputs:/workspace/outputs \
    -e KERAS_SOURCE="git+https://github.com/JyotinderSingh/keras@awq-2" \
    -e KERAS_HUB_SOURCE="git+https://github.com/JyotinderSingh/keras-hub@awq" \
    jyotindersingh/awq-benchmark benchmark \
    --model-class keras_hub.models.Gemma3CausalLM \
    --model-preset gemma3_1b
```

See [Keras-AWQ-Benchmark Repository](https://github.com/JyotinderSingh/Keras-AWQ-Benchmark) for details.

---

awq_benchmark_analysis

### CODE REVIEW

File: keras/src/quantizers/quantization_config.py [hunk 1]
```diff
@@ -220,6 +225,15 @@ def validate_and_resolve_config(mode, config):
                 f"`GPTQConfig`. Received: {type(config)}"
             )
 
+    if mode == "awq":
+        from keras.src.quantizers.awq_config import AWQConfig
```
Comment (hertschuh) [inline keras/src/quantizers/quantization_config.py:229]: We only do late imports when needed (i.e. circular import). Move this one to the top (and the `GPTQConfig` one too).

File: keras/src/models/model.py [hunk 2]
```diff
@@ -568,6 +568,25 @@ def quantize(self, mode=None, config=None, filters=None, **kwargs):
...
+
+            if structure is None:
+                raise ValueError(
+                    "For 'awq' mode, a valid quantization structure must be "
+                    "provided either via `config.quantization_layer_structure` "
+                    "or by overriding "
+                    "`model.get_quantization_layer_structure(mode)`. The "
+                    "structure should be a dictionary with keys "
+                    "'pre_block_layers' and 'sequential_blocks'."
+                )
```
Comment (hertschuh) [inline keras/src/models/model.py]: Share this code with the one above to avoid full duplication:
```
if mode in ("awq", "gptq"):
    ....
    ValueError(
       f"  {mode}  "
    if mode == "awq":
        ....
```

File: keras/src/layers/core/einsum_dense.py [hunk 3]
```diff
@@ -616,6 +637,125 @@ def _gptq_call(self, inputs, training=False):
...
+
+        # Ensures the forward pass uses the original high-precision kernel
+        # until calibration has been performed.
+        self.is_awq_calibrated = False
+
+        self.original_kernel_shape = kernel_shape
+        if len(kernel_shape) == 2:
+            rows = kernel_shape[0]
+            columns = kernel_shape[1]
+        elif len(kernel_shape) == 3:
```
Comment (hertschuh) [inline keras/src/layers/core/einsum_dense.py:656]: Should you add an `else: raise ValueError` at the end explaining that 1D and 4D and up is not supported?

File: keras/src/quantizers/quantizers.py [hunk 4]
```diff
@@ -653,11 +653,14 @@ def unpack_int4(packed, orig_len, axis=0, dtype="int8"):
         )
 
     def to_signed(x):
-        """Converts unpacked nibbles [0, 15] to signed int4 [-8, 7]."""
+        """Converts unpacked nibbles [0, 15] to signed int4 [-8, 7].
```
Comment (hertschuh) [inline keras/src/quantizers/quantizers.py:656]: It would be nice to add `int4` somewhere in the name of the function.

**Review by hertschuh — COMMENTED** [review]
Thanks for adding this! Some small comments:

File: keras/src/quantizers/quantization_config.py (hunk 1 above)
Comment (JyotinderSingh) [reply keras/src/quantizers/quantization_config.py:229]: `AWQConfig` (defined in awq_config.py) and `GPTQConfig` (defined in gtpq_config.py) depend on QuantizationConfig (defined in this file), and the `validate_and_resolve_config` utility (this method) makes use of `AWQConfig` and `GPTQConfig` for type checks - this causes a circular import, for which I've added the lazy import.

I can work on fixing some of these circular imports in a separate change (since they exist in dense.py/einsum_dense.py, etc. as well)

File: keras/src/models/model.py (hunk 2 above)
Comment (JyotinderSingh) [reply keras/src/models/model.py]: Done.

File: keras/src/layers/core/einsum_dense.py (hunk 3 above)
Comment (JyotinderSingh) [reply keras/src/layers/core/einsum_dense.py:656]: Done.

File: keras/src/quantizers/quantizers.py (hunk 4 above)
Comment (JyotinderSingh) [reply keras/src/quantizers/quantizers.py:656]: Renamed the method to `convert_to_signed_int4()`

**Review by hertschuh — APPROVED** [review]
